File: netapp-manageability-sdk-9.3/src/sample/Data_ONTAP/Cluster-Mode/Perl/vserverlist.pl
#============================================================#
#                                                            #
#                                                            #
# vserverlist.pl                                             #
#                                                            #
# Sample code to list the vservers available in the cluster. #
# This sample code demonstrates the usage of new iterative   #
# vserver-get-iter API.                                      #
# This sample code is supported from Cluster-Mode            #
# Data ONTAP 8.1 onwards.                                    #
#                                                            #
# Copyright 2011 NetApp, Inc. All rights reserved.           #
# Specifications subject to change without notice.           #
#                                                            #
#============================================================#

use lib "../../../../../lib/perl/NetApp"; 
use NaServer;
use NaElement;
use strict;

my $args = $#ARGV + 1;
my ($server, $ipaddr, $user, $passwd);


sub print_usage_and_exit() {
    print("\nUsage: \n");
    print("vserverlist <cluster/vserver> <user> <passwd> [-v <vserver-name>] \n");
    print("<cluster>             -- IP address of the cluster \n");
    print("<vserver>             -- IP address of the vserver \n");
    print("<user>                -- User name \n");
    print("<passwd>              -- Password \n");
    print("<vserver-name>        -- Name of the vserver \n\n");
    print("Note: \n");
    print(" -v switch is required when you want to tunnel the command to a vserver using cluster interface \n\n");
    exit (-1);
}

sub list_vservers {
   my ($in, $out);
    my ($rootVol, $rootVolAggr, $secStyle, $state);
    my $tag = "";

   if ($args > 3) {
        if ($args < 5 || $ARGV[3] ne ("-v")) {
            print_usage_and_exit();
        }
        $server->set_vserver($ARGV[4]);
    }
    while (defined($tag)) {
        $in = NaElement->new("vserver-get-iter");
        if ($tag ne "") {
            $in->child_add_string("tag", $tag);
        }
        $out = $server->invoke_elem($in);
        if($out->results_status() eq "failed") {
            print($out->results_reason() ."\n");
            exit(-1);
        }
        if ($out->child_get_int("num-records") == 0) {
            print("No vserver(s) information available \n");
            return;
        }
        $tag = $out->child_get_string("next-tag");
        my @vserverList = $out->child_get("attributes-list")->children_get();
        print("----------------------------------------------------\n");
        my $vserverInfo;
        # Iterate through each vserver info
        foreach $vserverInfo (@vserverList) {
            print("Name                    : " . $vserverInfo->child_get_string("vserver-name") . "\n");
            print("Type                    : " . $vserverInfo->child_get_string("vserver-type") . "\n");
            $rootVolAggr = $vserverInfo->child_get_string("root-volume-aggregate");
            $rootVol = $vserverInfo->child_get_string("root-volume");
            $secStyle = $vserverInfo->child_get_string("root-volume-security-style");
            $state = $vserverInfo->child_get_string("state");
            print("Root volume aggr        : " . ($rootVolAggr ? $rootVolAggr : "") . "\n");
            print("Root volume             : " . ($rootVol ? $rootVol : "") . "\n");
            print("Root volume sec style   : " . ($secStyle ? $secStyle : "") . "\n");
            print("UUID                    : " . $vserverInfo->child_get_string("uuid") . "\n");
            print("State                   : " . ($state ? $state : "") . "\n");
            my $allowedProtocols;
            print("Allowed protocols       : ");
            if (($allowedProtocols = $vserverInfo->child_get("allowed-protocols"))) {
                my @allowedProtocolsList = $allowedProtocols->children_get();
                foreach my $protocol (@allowedProtocolsList) {
                   print($protocol->get_content() . " ");
                }
            }
            print("\nName server switch      : ");
            my $nameServerSwitch;
            if (($nameServerSwitch = $vserverInfo->child_get("name-server-switch"))) {
                my @nsSwitchList = $nameServerSwitch->children_get();
                foreach my $nsSwitch (@nsSwitchList) {
                    print($nsSwitch->get_content() . " ");
                }
            }
            print("\n----------------------------------------------------\n");
        }
    }
}

sub main() {

    if ($args < 3) {
        print_usage_and_exit();
    }
    $ipaddr = $ARGV[0];
    $user = $ARGV[1];
    $passwd = $ARGV[2];

    $server = NaServer->new($ipaddr, 1, 15);
    $server->set_style("LOGIN");
    $server->set_admin_user($user, $passwd);
    $server->set_transport_type("HTTP");
    list_vservers();
}

main();

File: netapp-manageability-sdk-9.3/src/sample/Data_ONTAP/Cluster-Mode/DotNet/VB.NET/vserverlist/vserverlist.vb
'============================================================'
'                                                            '
' vserverlist.vb                                             '
'                                                            '
' Sample code to list the volumes available in the cluster.  '
'                                                            '
' This sample code is supported from Cluster-Mode            '
' Data ONTAP 8.1 onwards.                                    '
'                                                            '
' Copyright 2011 NetApp, Inc. All rights reserved.           '
' Specifications subject to change without notice.           '
'                                                            '
'============================================================'

Imports System
Imports System.Collections.Generic
Imports NetApp.Manage

Module VserverList
    Dim server As NaServer
    Dim args As String()

    Public Sub PrintUsageAndExit()
        Console.WriteLine(Environment.NewLine & "Usage:")
        Console.WriteLine("vserverlist <cluster/vserver> <user> <passwd> [-v <vserver-name>]")
        Console.WriteLine("<cluster>             -- IP address of the cluster")
        Console.WriteLine("<vserver>             -- IP address of the vserver")
        Console.WriteLine("<user>                -- User name")
        Console.WriteLine("<passwd>              -- Password")
        Console.WriteLine("<vserver-name>        -- Name of the vserver")
        Console.WriteLine("Note: ")
        Console.WriteLine(" -v switch is required when you want to tunnel the command to a vserver using cluster interface")
        Environment.Exit(-1)
    End Sub

    Sub ListVolumes()
        Dim xi, xo As NaElement
        Dim tag As String = ""
        Dim flag As Boolean = True

        While (flag = True)
            If (args.Length > 4) Then
                If (args.Length < 6 Or (Not args(4).Equals("-v"))) Then
                    PrintUsageAndExit()
                End If
                server.Vserver = args(5)
            End If
            xi = New NaElement("vserver-get-iter")
            If (Not tag.Equals("")) Then
                xi.AddNewChild("tag", tag)
            End If
            xo = server.InvokeElem(xi)
            If (xo.GetChildIntValue("num-records", 0) = 0) Then
                Console.WriteLine("No vserver(s) information available\n")
                Return
            End If
            tag = xo.GetChildContent("next-tag")
            If (tag = Nothing) Then
                flag = False
            End If
            Dim vserverList As List(Of NaElement) = xo.GetChildByName("attributes-list").GetChildren()
            Dim total As Integer = vserverList.Count
            Dim index As Integer
            Dim vserverInfo As NaElement
            Dim protocol As NaElement
            Dim nsSwitch As NaElement
            Dim rootVolAggr As String
            Dim rootVol As String
            Dim secStyle As String
            Dim state As String

            Console.WriteLine("----------------------------------------------------")
            For index = 0 To total - 1
                vserverInfo = vserverList(index)
                Console.WriteLine("Name                    : " + vserverInfo.GetChildContent("vserver-name"))
                Console.WriteLine("Type                    : " + vserverInfo.GetChildContent("vserver-type"))
                rootVolAggr = vserverInfo.GetChildContent("root-volume-aggregate")
                rootVol = vserverInfo.GetChildContent("root-volume")
                secStyle = vserverInfo.GetChildContent("root-volume-security-style")
                state = vserverInfo.GetChildContent("state")
                Console.WriteLine("Root volume aggregate   : " + IIf(rootVolAggr <> Nothing, rootVolAggr, ""))
                Console.WriteLine("Root volume             : " + IIf(rootVol <> Nothing, rootVol, ""))
                Console.WriteLine("Root volume sec style   : " + IIf(secStyle <> Nothing, secStyle, ""))
                Console.WriteLine("UUID                    : " + vserverInfo.GetChildContent("uuid"))
                Console.WriteLine("State                   : " + IIf(state <> Nothing, state, ""))
                Dim allowedProtocols As NaElement = vserverInfo.GetChildByName("allowed-protocols")
                Console.Write("Allowed protocols       : ")
                If (Not (allowedProtocols Is Nothing)) Then
                    Dim allowedProtocolsList As List(Of NaElement) = allowedProtocols.GetChildren()
                    Dim protoIndex As Integer
                    For protoIndex = 0 To allowedProtocolsList.Count - 1
                        protocol = allowedProtocolsList(protoIndex)
                        Console.Write(protocol.GetContent() + " ")
                    Next protoIndex
                End If
                Dim nameServerSwitch As NaElement = vserverInfo.GetChildByName("name-server-switch")
                Console.Write(Environment.NewLine & "Name server switch      : ")
                If (Not (nameServerSwitch Is Nothing)) Then
                    Dim nameServerSwitchList As List(Of NaElement) = nameServerSwitch.GetChildren()
                    Dim nsSwitchIndex As Integer
                    For nsSwitchIndex = 0 To nameServerSwitchList.Count - 1
                        nsSwitch = nameServerSwitchList(nsSwitchIndex)
                        Console.Write(nsSwitch.GetContent() + " ")
                    Next nsSwitchIndex
                End If
                Console.WriteLine(Environment.NewLine & "----------------------------------------------------")
            Next index
        End While
    End Sub

    Sub Main()
        args = Environment.GetCommandLineArgs()

        If (args.Length < 4) Then
            PrintUsageAndExit()
            Exit Sub
        End If

        Dim serverName As String = args(1)
        Dim user As String = args(2)
        Dim passwd As String = args(3)

        Try
            'Initialize connection to server, and request version 1.15 of the API set.
            server = New NaServer(serverName, 1, 15)

            server.ServerType = NaServer.SERVER_TYPE.FILER
            server.TransportType = NaServer.TRANSPORT_TYPE.HTTP
            server.SetAdminUser(user, passwd)

            ListVolumes()

        Catch ex As NaException
            Console.WriteLine(ex.Message)
        Catch ex As Exception
            Console.WriteLine(ex.Message)
        End Try
    End Sub

End Module
